pkg/apiserver/domain/service/pipeline.go

```diff
@@ -110,90 +122,195 @@ func NewContextService() ContextService {
+func (p pipelineServiceImpl) GetPipeline(ctx context.Context, name string, getInfo bool) (*apis.GetPipelineResponse, error) {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	pipeline := &model.Pipeline{
+		Name:    name,
+		Project: project.Name,
+	}
+	if err := p.Store.Get(ctx, pipeline); err != nil {
+		if errors.Is(err, datastore.ErrRecordNotExist) {
+			return nil, bcode.ErrPipelineNotExist
+		}
 		return nil, err
 	}
+	base := pipeline2PipelineBase(pipeline, *project)
+	var info = apis.PipelineInfo{}
+	if getInfo {
+		in, err := p.getPipelineInfo(ctx, pipeline, project.Namespace)
+		if err != nil {
+			log.Logger.Errorf("get pipeline %s/%s info error: %v", pipeline.Project, pipeline.Name, err)
+			return nil, bcode.ErrGetPipelineInfo
+		}
+		if in != nil {
+			info = *in
+		}
+	}
+
 	return &apis.GetPipelineResponse{
-		PipelineBase: *workflow2PipelineBase(wf),
-		// todo info
+		PipelineBase: *base,
+		PipelineInfo: info,
 	}, nil
 }
 
 // UpdatePipeline will update a pipeline
-func (p pipelineServiceImpl) UpdatePipeline(ctx context.Context, name, project string, req apis.UpdatePipelineRequest) (*apis.PipelineBase, error) {
-	wf := v1alpha1.Workflow{}
-	if err := p.KubeClient.Get(ctx, client.ObjectKey{Name: name, Namespace: nsForProj(project)}, &wf); err != nil {
+func (p pipelineServiceImpl) UpdatePipeline(ctx context.Context, name string, req apis.UpdatePipelineRequest) (*apis.PipelineBase, error) {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	if err := checkPipelineSpec(req.Spec); err != nil {
 		return nil, err
 	}
-	wf.WorkflowSpec = req.Spec
-	wf.SetLabels(map[string]string{
-		labelDescription: req.Description,
-		labelAlias:       req.Alias})
-	if err := p.KubeClient.Update(ctx, &wf); err != nil {
+	pipeline := &model.Pipeline{
+		Name:    name,
+		Project: project.Name,
+	}
+	if err := p.Store.Get(ctx, pipeline); err != nil {
+		if errors.Is(err, datastore.ErrRecordNotExist) {
+			return nil, bcode.ErrPipelineNotExist
+		}
+		return nil, err
+	}
+
+	pipeline.Spec = req.Spec
+	pipeline.Description = req.Description
+	pipeline.Alias = req.Alias
+
+	if err := p.Store.Put(ctx, pipeline); err != nil {
 		return nil, err
 	}
-	return workflow2PipelineBase(wf), nil
+	return pipeline2PipelineBase(pipeline, *project), nil
 }
 
 // DeletePipeline will delete a pipeline
 func (p pipelineServiceImpl) DeletePipeline(ctx context.Context, pl apis.PipelineBase) error {
-	wf := v1alpha1.Workflow{}
-	if err := p.KubeClient.Get(ctx, client.ObjectKey{Name: pl.Name, Namespace: nsForProj(pl.Project)}, &wf); err != nil {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	pipeline := &model.Pipeline{
+		Name:    pl.Name,
+		Project: project.Name,
+	}
+	if err := p.Store.Get(ctx, pipeline); err != nil {
+		if errors.Is(err, datastore.ErrRecordNotExist) {
+			return bcode.ErrPipelineNotExist
+		}
+		return err
+	}
+	if err := p.Store.Delete(ctx, pipeline); err != nil {
```

Comment:
Delete the pipeline after the run deletion?

Reply:
change the order to run->context->pipeline